src/Workspaces/Remote/ServiceHub/Host/RemoteAnalyzerAssemblyLoaderService.cs

```diff
@@ -16,16 +18,52 @@ namespace Microsoft.CodeAnalysis.Remote.Diagnostics
     [ExportWorkspaceService(typeof(IAnalyzerAssemblyLoaderProvider), WorkspaceKind.RemoteWorkspace), Shared]
     internal sealed class RemoteAnalyzerAssemblyLoaderService : IAnalyzerAssemblyLoaderProvider
     {
-        private readonly DefaultAnalyzerAssemblyLoader _loader = new DefaultAnalyzerAssemblyLoader();
-        private readonly ShadowCopyAnalyzerAssemblyLoader _shadowCopyLoader = new ShadowCopyAnalyzerAssemblyLoader(Path.Combine(Path.GetTempPath(), "VS", "AnalyzerAssemblyLoader"));
+        private readonly RemoteAnalyzerAssemblyLoader _loader;
+        private readonly ShadowCopyAnalyzerAssemblyLoader _shadowCopyLoader;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public RemoteAnalyzerAssemblyLoaderService()
         {
+            _loader = new(Path.GetDirectoryName(Path.GetFullPath(typeof(RemoteAnalyzerAssemblyLoader).GetTypeInfo().Assembly.Location)));
+            _shadowCopyLoader = new(Path.Combine(Path.GetTempPath(), "VS", "AnalyzerAssemblyLoader"));
         }
 
         public IAnalyzerAssemblyLoader GetLoader(in AnalyzerAssemblyLoaderOptions options)
             => options.ShadowCopy ? _shadowCopyLoader : _loader;
+
+        // For analyzers shipped in Roslyn, different set of assemblies might be used when running
+        // in-proc and OOP e.g. in-proc (VS) running on desktop clr and OOP running on ServiceHub .Net6
+        // host. We need to make sure to use the right one for OOP.
+        private class RemoteAnalyzerAssemblyLoader : DefaultAnalyzerAssemblyLoader
+        {
+            private readonly string? _baseDirectory;
+
+            protected override Assembly LoadImpl(string fullPath)
+                => base.LoadImpl(FixPath(fullPath));
+
+            public RemoteAnalyzerAssemblyLoader(string? baseDirectory)
+            {
+                _baseDirectory = baseDirectory;
+            }
+
+            private string FixPath(string fullPath)
+            {
+                if (_baseDirectory == null)
+                {
+                    return fullPath;
+                }
+
+                var assemblyName = Path.GetFileName(fullPath);
+                var fixedPath = Path.GetFullPath(Path.Combine(_baseDirectory, assemblyName));
+
+                if (!PathUtilities.PathsEqual(fixedPath, Path.GetFullPath(fullPath)) && File.Exists(fixedPath))
```

Comment:
>!PathUtilities.PathsEqual(fixedPath, Path.GetFullPath(fullPath)) [](http://example.com/codeflow?start=20&length=64)

Why compare equality? If they are equal then it doesn't matter which one is returned, right?

Reply:
yes, this is just to shortcut the FileExists call

Reply:
Since the assembly is gonna be read from disk anyways I don't think avoiding File.Exists has any benefit

Reply:
Fixed